src/axom/core/Array.hpp

```diff
@@ -120,10 +105,20 @@ class Array : public ArrayBase<T, DIM, Array<T, DIM>>
    * \post size() == num_elements
    * \post getResizeRatio() == DEFAULT_RESIZE_RATIO
    */
-  template <IndexType SFINAE = DIM, typename std::enable_if<SFINAE == 1>::type* = nullptr>
+  template <IndexType SFINAE_DIM = DIM,
+            MemorySpace SFINAE_SPACE = SPACE,
+            typename std::enable_if<SFINAE_DIM == 1 &&
+                                    SFINAE_SPACE == MemorySpace::Dynamic>::type* = nullptr>
   Array(IndexType num_elements,
         IndexType capacity = 0,
-        int allocator_id = axom::getDefaultAllocatorID());
+        int allocator_id = axom::detail::getAllocatorID<SPACE>());
+
+  /// \overload
+  template <IndexType SFINAE_DIM = DIM,
+            MemorySpace SFINAE_SPACE = SPACE,
+            typename std::enable_if<SFINAE_DIM == 1 &&
+                                    SFINAE_SPACE != MemorySpace::Dynamic>::type* = nullptr>
```

Comment:
I am starting to get a little worried about relying on SFINAE in `axom::Array` from a maintainability standpoint.
- Do you have a sense of the compile time impact for instantiating `axom::Array`?
- How does SFINAE compare with alternative approaches, like putting these kinds of functions in a partially-specialized base class? Seems like alot of our functions are along the lines of "enable only for DIM == 1/DIM != 1"

(These don't have to be resolved/answered for this PR -- just something to consider)

Reply:
I was able to replace some of its uses (for conditionally enabled member functions) with `static_assert` - but I think it's pretty close to unavoidable for the three remaining uses in constructors, specifically:

- restricting the variadic ctor (`Args... args`) to only accept parameter packs of integral type - with C++20 I'd be able to use a concept, but even then I'm not sure what the difference in compile times would be 
- allowing an allocator ID only for the 1D `(int elements, int capacity)` constructor when `MemSpace == Dynamic` - this one is technically not required since I could just add a runtime warning like I do in the copy ctor (where SFINAE is not possible), but I'd prefer to be able to restrict that at compile time.